runtime/lua/vim/lsp.lua

```diff
@@ -1274,18 +1275,24 @@ function lsp.buf_request(bufnr, method, params, handler)
       if request_success then
         client_request_ids[client_id] = request_id
       end
+    else
+      table.insert(unsupported_clients, client_id)
     end
   end)
 
-  -- if has client but no clients support the given method, call the callback with the proper
-  -- error message.
+  -- if has clients but no clients support the given method, call the callback once for each unsupported client
+  -- with the proper error message.
   if not tbl_isempty(all_buffer_active_clients[resolve_bufnr(bufnr)] or {}) and not method_supported then
     local unsupported_err = lsp._unsupported_method(method)
     handler = handler or lsp.handlers[method]
     if handler then
-      handler(unsupported_err, method, bufnr)
+      vim.schedule(function()
+        for _, client_id in ipairs(unsupported_clients) do
+          handler(unsupported_err, method, nil, client_id, bufnr)
+        end
+      end)
+      return unsupported_clients
```

Comment:
Currently the return value is documented as:

```
--@returns 2-tuple:
---  - Map of client-id:request-id pairs for all successful requests.
---  - Function which can be used to cancel all the requests. You could instead
---    iterate all clients and call their `cancel_request()` methods.
```

I think to make the error result compatible and non-surprising we could change it to a 3 result tuple

```suggestion
      return {}, function() end, unsupported_clients
```

(This then also needs a docs update)

Reply:
Makes sense!